### ISSUE TITLE
unexpected expression '(Vector Rat) 28' - compilation to LossFunctions

### ISSUE THREAD

**Issue by ndslusarz** [vehicle-lang/vehicle#386]
When trying to compile to loss functions a modified mnists-spec (slightly simplifed to avoid other issues and with a different robustness not referring to label but rather difference in value between outputs) which can be found below:

```
-- Inputs and outputs

-- Define the type for our input images
--type Image = Tensor Rat [28, 28]

--------------------------------------------------------------------------------
-- Network

-- Declare the network used to classify images. The output of the network is a
-- score for each of the digits 0 to 9.
@network
mnist : Tensor Rat [28, 28] -> Vector Rat 10

--------------------------------------------------------------------------------
-- Definition of robustness around a point

-- First we define the parameter `epsilon` that will represent the radius of the
-- ball that we want the network to be robust in. Note that we declare this as
-- a parameter which allows the value of `epsilon` to be specified at compile
-- time rather than be fixed in the specification.

--epsilon : Rat
--epsilon = 0.01

--delta : Rat
--delta = 0.02

-- Next we define what it means for an image `x` to be in a ball of
-- size epsilon around 0.
boundedByEpsilon : Tensor Rat [28, 28] -> Bool
boundedByEpsilon x = forall i j . -0.01 <= x ! i ! j <= 0.01

--The disjunction avoids having to define absolute value
boundedByDelta : Tensor Rat [28, 28] -> Tensor Rat [28, 28] -> Bool
boundedByDelta x y = forall i . (-0.02 <= (((mnist x) ! i ) - ((mnist y) ! i )) <= 0.02) 
                                or (-0.02 <= (-((mnist x) ! i ) - ((mnist y) ! i )) <= 0.02)
                                or (-0.02 <= (((mnist x) ! i) + ((mnist y) ! i )) <= 0.02)
                                or (-0.02 <= (-((mnist x) ! i ) - ((mnist y) ! i )) <= 0.02)


robustSingleInput : Tensor Rat [28, 28] -> Bool
robustSingleInput image = forall pertubation .
    let perturbedImage = image - pertubation in
    boundedByEpsilon pertubation => boundedByDelta image perturbedImage

@property
robust1 : Bool
robust1 = forall x . robustSingleInput x

```
I get an error `encountered unexpected expression '(Vector Rat) 28' during compilation to loss functions.`

Through trial and error this seems to be caused by line `boundedByEpsilon pertubation => boundedByDelta image perturbedImage` although I'm unsure which part of it. The constraint type checks correctly and one dimensional vectors in other specs have not caused this issue before. I'm unsure whether there is an issue in this test spec somewhere or, more likely, if there is some problem with the loss function translation.

**Comment by ndslusarz** [comment]
Can't reproduce the error anymore on the current version with the same specification. However currently there is a typechecking error when trying to compile the same spec:

```
 "Error during specification compilation: " + result.stderr.decode("UTF-8")       
Exception: Error during specification compilation: Error at Line 44, Columns 32-33: unable to find a consistent type for the overloaded expression '-'. In particular 28 != image.
Fix: check your types
```

**Comment by wenkokke** [comment]
@MatthewDaggitt We should probably make that error message less hostile 😂

**Comment by MatthewDaggitt** [comment]
This was fixed in 3a3548f9b48c3969212ccb51e954d4d4556ea815
